dialdbg/src/parse.rs

```diff
@@ -0,0 +1,326 @@
+        if let Some(a) = self.mdns_address {
+            writeln!(f, "mDNS address {} was used for connection", a)?;
+        }
+        match self.mdns {
+            Some(d) => {
+                writeln!(f, "mDNS queried in {}ms", d.num_milliseconds(),)?;
+            }
+            None => {
+                writeln!(f, "mDNS could not be used to connect")?;
+            }
+        }
+
+        match self.authentication {
+            Some(d) => {
+                writeln!(f, "authentication successful in {}ms", d.num_milliseconds(),)?;
+            }
+            None => {
+                writeln!(f, "authentication failed")?;
+            }
+        }
+
+        match self.connection {
+            Some(d) => {
+                writeln!(
+                    f,
+                    "WebRTC connection establishment successful in {}ms",
+                    d.num_milliseconds(),
+                )?;
+            }
+            None => {
+                writeln!(f, "WebRTC connection establishment failed")?;
+            }
+        }
+
+        if let Some(emsg) = &self.dial_error_message {
+            writeln!(f, "\n{emsg}")?;
+        }
+
+        if let Some(c) = &self.selected_candidate_pair {
+            writeln!(f, "selected ICE candidate pair was:\n\t{c}")?;
+        }
+
+        Ok(())
+    }
+}
+
+fn extract_timestamp(log: &str) -> Result<DateTime<FixedOffset>> {
+    let split_log = log.split_whitespace().collect::<Vec<&str>>();
+    if split_log.len() == 0 {
+        return Err(anyhow!("malformed log returned by dial: {log}"));
+    }
+    match DateTime::parse_from_rfc3339(split_log[0]) {
+        Ok(d) => Ok(d),
+        Err(e) => Err(anyhow!("error parsing timestamp in log {log}: {e}")),
+    }
+}
+
+fn extract_mdns_address(log: &str) -> Result<SocketAddr> {
+    let mut split_log = log.split_whitespace().collect::<Vec<&str>>();
+
+    // mDNS IP address should be last token in log.
+    match split_log.pop() {
+        Some(a) => match a.parse::<SocketAddr>() {
+            Ok(a) => Ok(a),
+            Err(e) => Err(anyhow!("error parsing IP address {a} in log {log}: {e}",)),
+        },
+        None => Err(anyhow!("malformed mDNS log returned by dial: {log}")),
+    }
+}
+
+fn extract_dial_error(log: &str) -> Result<String> {
+    // Tear off LOG prefixes and reattach the DIAL_ERROR_PREFIX.
+    let split_log = log.split(strings::DIAL_ERROR_PREFIX).collect::<Vec<&str>>();
+    if split_log.len() != 2 {
+        return Err(anyhow!("malformed dial error message: {log}"));
+    }
+    Ok(format!("{}{}", strings::DIAL_ERROR_PREFIX, split_log[1]))
+}
+
+pub(crate) fn parse_grpc_logs(log_path: PathBuf, out: &mut Box<dyn io::Write>) -> Result<()> {
+    let mut res = GRPCResult::default();
+
+    let mut connection_establishment_start = None;
+    let mut authentication_start = None;
+    let mut mdns_query_start = None;
+    for log in fs::read_to_string(log_path)?.lines() {
```

Comment:
(q) it doesn't look to me like this validates around duplicate messages. Is this correct? If so, are there cases where we should be doing that validation?

The same for the `parse_webrtc_logs` function below.

Reply:
A great point. Right now we simply honor the last message of its kind in setting fields on result types. I don't think `rust-utils` would ever output duplicates of the messages we're interested in unless something was wrong with dial. I _could_ add validation, but it's not "trivial" (makes the code look a little messy). I think right now, the last-message-honored approach is probably ok, but I'm happy to add duplicate errors if you think it's more robust.

Reply:
I think it's fine to leave as is, thanks for clarifying! We can always add duplicate error checking down the line.